src/io/measure_record_reader.h

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef STIM_RECORD_READER_H
+#define STIM_RECORD_READER_H
+
+#include <memory>
+
+#include "../circuit/circuit.h"
+#include "../simd/pointer_range.h"
+#include "../simd/simd_bit_table.h"
+
+namespace stim_internal {
+
+/// Handles reading measurement data from the outside world.
+///
+/// Child classes implement the various input formats. Each file format encodes a certain number of records.
+/// Each record is a sequence of 0s and 1s. File formats B8 and R8 encode a single record. File formats 01,
+/// HITS and DETS encode any number of records. File formats 01, B8 and R8 encode the length of each record.
```

Comment:
I think the reader should *always* know how long the record is going to be. The length should be required in order to create the reader. The 01 and DETS and other length-in-data readers use this information for *verifying* that the number of records they get is correct. The reader should be responsible for length validation; not the caller.